rng/philox/philox.hpp

```diff
@@ -0,0 +1,202 @@
+
+    // equality operators
+    friend bool operator==(const philox_engine& x, const philox_engine& y) {
+        return false;
+    }
+
+    // generating functions
+    result_type operator()()
+    {
+        ++i;
+        if (i == n) {
+            y = philox_generate(k, x); // see below
+            increment_counter();
+            i = 0;
+        }
+        return y[i];
+    }
+
+    void discard(unsigned long long z);
+    // inserters and extractors
+    template<class charT, class traits>
+    friend basic_ostream<charT, traits>&
+    operator<<(basic_ostream<charT, traits>& os, const philox_engine& x);
+
+    template<class charT, class traits>
+    friend basic_istream<charT, traits>&
+    operator>>(basic_istream<charT, traits>& is, philox_engine& x);
+
+
+private: // utilities
+    using uint_types = std::tuple<std::uint8_t, std::uint16_t, std::uint32_t, std::uint64_t>;
+    using promotion_types = std::tuple<std::uint16_t, std::uint32_t, std::uint64_t, __uint128_t>;
+
+    static consteval std::size_t log2(std::size_t val)
+    {
+        return ((val <= 2) ? 1 : 1 + log2(val / 2));
+    }
+
+    static consteval std::size_t ceil_log2(std::size_t val)
+    {
+        std::size_t additive = static_cast<std::size_t>(!std::has_single_bit(val));
+
+        return log2(val) + additive;
+    }
+
+    using counter_type = std::tuple_element_t<ceil_log2(w / CHAR_BIT), uint_types>;
+    using promotion_type = std::tuple_element_t<ceil_log2(w / CHAR_BIT), promotion_types>;
+
+    static constexpr counter_type counter_mask = ~counter_type(0) >> (sizeof(counter_type) * CHAR_BIT - w);
+    static constexpr result_type result_mask = ~result_type(0) >> (sizeof(result_type) * CHAR_BIT - w);
+    
+
+private: // functions
+
+    static std::pair<counter_type, counter_type> mulhilo(result_type a, result_type b)
+    {
+        constexpr std::size_t shift = std::numeric_limits<promotion_type>::digits - w;
+        promotion_type promoted_a = a;
+        promotion_type promoted_b = b;
+        promotion_type result = promoted_a * promoted_b;
+        counter_type mulhi = result >> shift;
+        counter_type mullo = (result << shift) >> shift;
+        return {mulhi, mullo};
+    }
+
+    static std::array<result_type, n> philox_generate(std::array<result_type, array_size> keys, std::array<counter_type, n> x)
+    {
+        for (std::size_t q = 0; q != r; ++q)
+        {
+            std::array<counter_type, n> v = x;
+            if constexpr(n == 4)
+            {
+                v[0] ^= v[2] ^= v[0] ^= v[2];
+            }
+            for (std::size_t k = 0; k < array_size; ++k)
+            {
+                auto [mulhi, mullo] = mulhilo(v[2 * k], multipliers[k]);
+                x[2 * k + 1] = mullo;
+                x[2 * k] = mulhi ^ keys[k] ^ v[2 * k + 1];
+                keys[k] = (keys[k] + round_consts[k]) & result_mask;
+            }
+        }
+        return [&x]<std::size_t... Is>(std::index_sequence<Is...>) {
```

Comment:
C++ developers do not look for simple solution... And when you are using hammer for too long, everything looks like a nail...